File: typescript/src/integrations/openai.ts
/**
 * neuromem/integrations/openai
 * ----------------------------
 * Drop-in context-aware wrapper around the OpenAI Chat Completions API.
 *
 * @example
 * ```ts
 * import OpenAI from "openai";
 * import { ContextAwareOpenAI } from "neuromem/integrations/openai";
 *
 * const client = new ContextAwareOpenAI({
 *   openaiClient: new OpenAI(),
 *   tokenBudget: 8000,
 *   model: "gpt-4o",
 *   systemPrompt: "You are a helpful assistant.",
 * });
 *
 * const reply = await client.chat("Tell me about relativity.");
 * console.log(reply);
 * console.log(client.stats());
 * ```
 */

import { ContextManager, ContextManagerOptions } from "../contextManager.js";
import { Summarizer } from "../summarizer.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Minimal interface matching OpenAI SDK's ChatCompletion client */
export interface OpenAILike {
  chat: {
    completions: {
      create(params: Record<string, unknown>): Promise<{
        choices: Array<{
          message: { content: string | null };
          delta?: { content?: string | null };
        }>;
      }>;
    };
  };
}

export interface ContextAwareOpenAIOptions {
  openaiClient: OpenAILike;
  model?: string;
  tokenBudget?: number;
  systemPrompt?: string;
  /** "extractive" | "abstractive". Default: "extractive" */
  summarizeMode?: "extractive" | "abstractive";
  contextManager?: ContextManager;
}

export interface ChatOptions {
  model?: string;
  temperature?: number;
  maxTokens?: number;
  extraParams?: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// ContextAwareOpenAI
// ---------------------------------------------------------------------------

export class ContextAwareOpenAI {
  private _client: OpenAILike;
  readonly model: string;
  readonly context: ContextManager;

  constructor(options: ContextAwareOpenAIOptions) {
    this._client = options.openaiClient;
    this.model = options.model ?? "gpt-4o-mini";

    if (options.contextManager) {
      this.context = options.contextManager;
    } else {
      const summarizer = new Summarizer({
        mode: options.summarizeMode ?? "extractive",
        client:
          options.summarizeMode === "abstractive"
            ? (options.openaiClient as any)
            : undefined,
        model: this.model,
      });

      this.context = new ContextManager({
        tokenBudget: options.tokenBudget ?? 4096,
        summarizer,
      });
    }

    if (options.systemPrompt) {
      // synchronously add (no await needed for system message — won't trigger prune alone)
      void this.context.addSystem(options.systemPrompt);
    }
  }

  // ------------------------------------------------------------------

  async chat(
    userMessage: string,
    options: ChatOptions = {},
  ): Promise<string> {
    await this.context.addUser(userMessage);
    const messages = await this.context.getMessages();

    const params: Record<string, unknown> = {
      model: options.model ?? this.model,
      messages,
      temperature: options.temperature ?? 0.7,
      max_tokens: options.maxTokens ?? 1024,
      ...(options.extraParams ?? {}),
    };

    const response = await this._client.chat.completions.create(params);
    const reply = response.choices[0].message.content ?? "";

    await this.context.addAssistant(reply);
    return reply;
  }

  // ------------------------------------------------------------------

  async setSystem(prompt: string): Promise<void> {
    this.context.replaceSystem(prompt);
  }

  async reset(keepSystem = true): Promise<void> {
    this.context.clear(keepSystem);
  }

  stats() {
    return this.context.stats();
  }

  /** Direct pass-through to the underlying OpenAI client */
  async rawCreate(params: Record<string, unknown>) {
    return this._client.chat.completions.create(params);
  }
}

File: neuromem/integrations/openai.py
"""
neuromem.integrations.openai
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Drop-in context-aware wrapper around the OpenAI Chat Completions API.

Usage::

    import openai
    from neuromem.integrations.openai import ContextAwareOpenAI

    client = ContextAwareOpenAI(
        openai_client=openai.OpenAI(),
        token_budget=8000,
        model="gpt-4o",
    )

    # Exactly like the standard API — but your context is auto-managed
    reply = client.chat("Tell me about relativity.")
    reply = client.chat("Can you give me an example?")   # history preserved

    print(client.context.stats())
"""

from __future__ import annotations

from typing import Any, Dict, Iterator, List, Optional

from ..context_manager import ContextManager
from ..scorer import MessageScorer
from ..summarizer import Summarizer
from ..pruner import Pruner


class ContextAwareOpenAI:
    """
    Wraps an ``openai.OpenAI`` client and manages the context window automatically.

    Parameters
    ----------
    openai_client :
        A pre-configured ``openai.OpenAI()`` instance.
    model : str
        Default model to use for chat completions.
    token_budget : int
        Maximum tokens for the context window.
    system_prompt : str, optional
        Initial system message inserted on construction.
    summarize_mode : "extractive" | "abstractive"
        How to summarize old context.  Use "abstractive" for better quality
        (requires the LLM to be called for summarization).
    context_manager : ContextManager, optional
        Supply your own pre-configured ContextManager to override all defaults.
    """

    def __init__(
        self,
        openai_client,
        model: str = "gpt-4o-mini",
        token_budget: int = 4096,
        system_prompt: Optional[str] = None,
        summarize_mode: str = "extractive",
        context_manager: Optional[ContextManager] = None,
    ) -> None:
        self._client = openai_client
        self.model = model

        if context_manager is not None:
            self.context = context_manager
        else:
            summarizer = Summarizer(
                mode=summarize_mode,
                client=openai_client if summarize_mode == "abstractive" else None,
                model=model,
            )
            self.context = ContextManager(
                token_budget=token_budget,
                summarizer=summarizer,
            )

        if system_prompt:
            self.context.add_system(system_prompt)

    # ------------------------------------------------------------------

    def chat(
        self,
        user_message: str,
        *,
        model: Optional[str] = None,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        stream: bool = False,
        extra_params: Optional[Dict[str, Any]] = None,
    ) -> str:
        """
        Send *user_message*, manage context automatically, return the reply.

        Parameters
        ----------
        user_message : str
        model : str, optional  — overrides instance default
        temperature : float
        max_tokens : int
        stream : bool          — if True, streams and returns concatenated text
        extra_params : dict    — additional kwargs forwarded to the API

        Returns
        -------
        str — the assistant's reply content
        """
        self.context.add_user(user_message)
        messages = self.context.get_messages()

        kwargs: Dict[str, Any] = dict(
            model=model or self.model,
            messages=messages,
            temperature=temperature,
            max_tokens=max_tokens,
            **(extra_params or {}),
        )

        if stream:
            reply = self._stream(kwargs)
        else:
            response = self._client.chat.completions.create(**kwargs)
            reply = response.choices[0].message.content or ""

        self.context.add_assistant(reply)
        return reply

    def _stream(self, kwargs: dict) -> str:
        kwargs["stream"] = True
        chunks: List[str] = []
        for chunk in self._client.chat.completions.create(**kwargs):
            delta = chunk.choices[0].delta.content or ""
            chunks.append(delta)
        return "".join(chunks)

    # ------------------------------------------------------------------

    def set_system(self, prompt: str) -> None:
        """Replace or set the system prompt."""
        self.context.replace_system(prompt)

    def reset(self, keep_system: bool = True) -> None:
        """Clear conversation history."""
        self.context.clear(keep_system=keep_system)

    def stats(self) -> dict:
        """Return context statistics."""
        return self.context.stats()

    # ------------------------------------------------------------------
    # Raw API pass-through (for advanced usage)
    # ------------------------------------------------------------------

    def raw_create(self, **kwargs) -> Any:
        """
        Direct pass-through to ``openai_client.chat.completions.create``.
        The caller is responsible for message management.
        """
        return self._client.chat.completions.create(**kwargs)
